src/Features/Core/Portable/PullMemberUp/MembersPuller.cs

```diff
@@ -0,0 +1,414 @@
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeGeneration;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.PullMemberUp;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Roslyn.Utilities;
+using static Microsoft.CodeAnalysis.CodeActions.CodeAction;
+using EditorMap = System.Collections.Immutable.ImmutableDictionary<Microsoft.CodeAnalysis.SyntaxTree, Microsoft.CodeAnalysis.Editing.DocumentEditor>;
+using SyntaxMap = System.Collections.Immutable.ImmutableDictionary<Microsoft.CodeAnalysis.ISymbol, Microsoft.CodeAnalysis.SyntaxNode[]>;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings.PullMemberUp
+{
+    internal class MembersPuller
+    {
+        internal static readonly MembersPuller Instance = new MembersPuller();
+
+        private MembersPuller()
+        {
+        }
+
+        /// <summary>
+        /// Return the CodeAction to pull selectedMember up to destinationType. If the pulling will cause error,
+        /// it will return null.
+        /// </summary>
+        internal CodeAction TryComputeCodeAction(
+            Document document,
+            ISymbol selectedMember,
+            INamedTypeSymbol destinationType)
+        {
+            var result = PullMembersUpAnalysisBuilder.BuildAnalysisResult(destinationType, ImmutableArray.Create((member: selectedMember, makeAbstract: false)));
+            if (result.PullUpOperationCausesError ||
+                IsSelectedMemberDeclarationAlreadyInDestination(destinationType, selectedMember))
+            {
+                return default;
+            }
+
+            return new SolutionChangeAction(
+                string.Format(FeaturesResources.Add_to_0, result.Destination),
+                cancellationToken => PullMembersUpAsync(result, document, cancellationToken));
+        }
+
+        /// <summary>
+        /// Return the changed solution if all changes in result are applied.
+        /// </summary>
+        /// <param name="result">Contains the members to pull up and all the fix operations</param>>
+        internal async Task<Solution> PullMembersUpAsync(
+            PullMembersUpAnalysisResult result,
+            Document contextDocument,
+            CancellationToken cancellationToken)
+        {
+            var solution = contextDocument.Project.Solution;
+            var solutionEditor = new SolutionEditor(solution);
+            var codeGenerationService = contextDocument.Project.LanguageServices.GetRequiredService<ICodeGenerationService>();
+            var destinationSyntaxNode = await codeGenerationService.FindMostRelevantNameSpaceOrTypeDeclarationAsync(
+                contextDocument.Project.Solution, result.Destination, default, cancellationToken).ConfigureAwait(false);
+            var (editorMap, syntaxMap) = await InitializeEditorMapsAndSyntaxMapAsync(result, destinationSyntaxNode, solutionEditor, solution, cancellationToken).ConfigureAwait(false);
```

Comment:
this code doesn't seem to be needed here.  It seems like it could just be created in PullMembersIntoInterface and PullMembersIntoClass